tokenbucket.py

```diff
@@ -0,0 +1,38 @@
+from time import time
+from threading import Lock
+
+
+class TokenBucket:
+    """
+    An implementation of the token bucket algorithm.
+    """
+    def __init__(self):
+        self.tokens = 0
+        self.rate = 0
+        self.last = time()
+        self.lock = Lock()
+
+    def set_rate(self, rate):
+        with self.lock:
+            self.rate = rate
+            self.tokens = self.rate
+
+    def consume(self, tokens):
+        with self.lock:
+            if not self.rate:
+                return 0
+
+            now = time()
+            lapse = now - self.last
+            self.last = now
+            self.tokens += lapse * self.rate
+
+            if self.tokens > self.rate:
+                self.tokens = self.rate
+
+            self.tokens -= tokens
+
+            if self.tokens >= 0:
+                return 0
+            else:
+                return -self.tokens / self.rate
```

Comment:
Tocken bucket is a pretty common algorithm. Isn't there a well known package?

Reply:
There's https://github.com/falconry/token-bucket on PyPy but a) it doesn't use thread locking and b) it's pretty much the same implementation that I've included, but with very verbose comments. I generally prefer to use existing packages too since it's one less thing that I need to maintain, but this was simple enough that I opted not to.